### PR TITLE
Expose `ClassDB.class_get_api_type()` method

### LINKED ISSUE #9526: Expose `ClassDB.get_api_type()` for better GDExtension interoperability with non-GDScript languages
### Describe the project you are working on

GDExtension is a great plugin system.
However, the current implementation of GDExtension as a plugin doesn't effectively communicate with languages other than GDScript (including other GDExtensions and C#). 
We aim to expose the ‘get_api_type’ method of ClassDB to allow developers to enumerate the types provided by GDExtension and generate corresponding bindings for other languages.

### Describe the problem or limitation you are having in your project

Currently it is almost impossible to easily obatin a list of currently activated GDExtension types from the ClassDB API.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

After exposing this method, the developer may use this API to fillter out the GDExtension types (instead of maintaining a list of hard-coded editor types).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```csharp
public static string[] GetGDExtensionTypes(bool includeEditorTypes = false)
    {
        var classList = ClassDB.GetClassList();
        var hash      = new HashSet<string>();
        foreach (var className in classList)
        {
            var apiType = ClassDB.GetApiType(className);
            if (apiType == ClassDB.ApiType.ApiExtesion)
            {
                hash.Add(className);
            }
            if (!includeEditorTypes)
            {
                continue;
            }
            if (apiType == ClassDB.ApiType.ApiEditorExtesion)
            {
                hash.Add(className);
            }
        }
        return hash.ToArray();
    }
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be achieve by manually record all Godot builtin types, and GodotTools builtin types by hand, for each Editor Release, which is neither easy nor sustainable for a wrapper generator.

### Is there a reason why this should be core and not an add-on in the asset library?

It is impossible to achieve the descirbed need without changing the Core.

### PR DESCRIPTION
**Pull request by ZerxZ**
Expose `ClassDB::class_get_api_type()` for editor-plugins and other languages wrapper generator.

- Closes https://github.com/godotengine/godot-proposals/issues/9526

### CODE REVIEW

**Review by dsnopek — COMMENTED** [review]
Thanks!

I think this change makes sense, and the code looks good.

However, this will need to be squashed into a single commit, per Godot's pull request workflow. See the docs for a way to do that using `git rebase`: https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase

I'll take a look at fixing the godot-cpp issue.

File: core/core_bind.h [hunk 1]
```diff
@@ -431,9 +431,19 @@ class ClassDB : public Object {
...
+		API_EXTENSION,
+		API_EDITOR_EXTENSION,
+		API_NONE,
+	};
+	APIType get_api_type(const StringName &p_class) const;
 	PackedStringArray get_class_list() const;
 	PackedStringArray get_inheriters_from_class(const StringName &p_class) const;
 	StringName get_parent_class(const StringName &p_class) const;
+	StringName get_property_setter(const StringName &p_class, const StringName &p_property) const;
+	StringName get_property_getter(const StringName &p_class, const StringName &p_property) const;
```
Comment (dsnopek) [inline core/core_bind.h:446]: Should the word `class` be in the method name somewhere?

I feel like all the other methods that are getting something about a particular class have it added, like:

- `get_integer_constant_enum` -> `class_get_integer_constant_enum`
- `is_enum_bitfield` -> `is_class_enum_bitfield`

Comment (dsnopek) [reply core/core_bind.h]: Thanks, this looks good to me!

**Comment by dsnopek** [comment]
> The `first_argument_is_class` variable in this code is becoming increasingly lengthy and less maintainable. Would it be more suitable to change it to a `HashSet<String>` field to store these names, and then use the `HashSet<String>` field to check for existence?

I would just leave this as it is for now.

Overall, this PR is looking good to me!

CI is complaining about an extra line in the docs XML:

```
diff --git a/doc/classes/ClassDB.xml b/doc/classes/ClassDB.xml
index ed670fa..6a58536 100644
--- a/doc/classes/ClassDB.xml
+++ b/doc/classes/ClassDB.xml
@@ -106,7 +106,6 @@
 				Returns the default value of [param property] of [param class] or its ancestor classes.
 			</description>
 		</method>
-
 		<method name="class_get_property_getter" qualifiers="const">
 			<return type="StringName" />
 			<param index="0" name="class" type="StringName" />
```

However, aside from that, the main thing blocking this is that we need to get godot-cpp PR https://github.com/godotengine/godot-cpp/pull/1445 merged and cherry-picked so that the rest of the tests can pass here.

**Comment by dsnopek** [comment]
Once PR https://github.com/godotengine/godot-cpp/pull/1465 is merged (which cherry-picks the necessary changes to `binding_generator.py`), then the tests here should finally start passing

**Comment by ZerxZ** [comment]
> Once PR [godotengine/godot-cpp#1465](https://github.com/godotengine/godot-cpp/pull/1465) is merged (which cherry-picks the necessary changes to `binding_generator.py`), then the tests here should finally start passing

Thanks!

**Review by dsnopek — APPROVED** [review]
The tests are now passing - yay!

This looks good to me, however, it would probably be good to update the PR title and commit message with the right method names.

File: core/core_bind.cpp [hunk 2]
```diff
@@ -1450,6 +1455,13 @@ Variant ClassDB::class_get_property_default_value(const StringName &p_class, con
 	}
 	return Variant();
 }
+StringName ClassDB::class_get_property_getter(const StringName &p_class, const StringName &p_property) const {
```
Comment (AThousandShips) [inline core/core_bind.cpp:1458]:
```suggestion

StringName ClassDB::class_get_property_getter(const StringName &p_class, const StringName &p_property) const {
```

File: core/core_bind.h [hunk 3]
```diff
@@ -431,14 +431,22 @@ class ClassDB : public Object {
...
+
 	PackedStringArray get_class_list() const;
 	PackedStringArray get_inheriters_from_class(const StringName &p_class) const;
 	StringName get_parent_class(const StringName &p_class) const;
 	bool class_exists(const StringName &p_class) const;
 	bool is_parent_class(const StringName &p_class, const StringName &p_inherits) const;
 	bool can_instantiate(const StringName &p_class) const;
 	Variant instantiate(const StringName &p_class) const;
-
+	APIType class_get_api_type(const StringName &p_class) const;
```
Comment (AThousandShips) [inline core/core_bind.h:449]:
```suggestion

	APIType class_get_api_type(const StringName &p_class) const;
```

**Review by AThousandShips — APPROVED** [review]

**Comment by Delsin-Yu** [comment]
Any chance we can have this in `4.3`? As the PR has approved.

**Comment by AThousandShips** [comment]
We're in feature freeze so I don't expect so, but we'll see what the production team says, I suspect 4.4

**Comment by dalexeev** [comment]
* See #94826.

I'm not sure about `class_get_api_type()`.

**Comment by ZerxZ** [comment]
> * See [Expose ClassDB methods `get_property_setter` / `get_property_getter` #94826](https://github.com/godotengine/godot/pull/94826).
> 
> I'm not sure about `class_get_api_type()`.

I think the class_get_api_type part is still necessary. It is mainly used to distinguish between Native Type and GDExtension Type.

**Comment by dalexeev** [comment]
> so I just need to keep the `class_get_api_type` one?

I think yes, since the other two methods are already exposed.

File: core/core_bind.cpp [hunk 4]
```diff
@@ -1609,7 +1614,8 @@ void ClassDB::get_argument_options(const StringName &p_function, int p_idx, List
 				pf == "class_has_method" || pf == "class_get_method_list" ||
 				pf == "class_get_integer_constant_list" || pf == "class_has_integer_constant" || pf == "class_get_integer_constant" ||
 				pf == "class_has_enum" || pf == "class_get_enum_list" || pf == "class_get_enum_constants" || pf == "class_get_integer_constant_enum" ||
-				pf == "is_class_enabled" || pf == "is_class_enum_bitfield");
+				pf == "is_class_enabled" || pf == "is_class_enum_bitfield" || pf == "class_get_api_type" ||
+				pf == "class_get_property_getter" || pf == "class_get_property_setter");
```
Comment (AThousandShips) [inline core/core_bind.cpp:1618]: Remove these two now they aren't exposed

File: core/core_bind.cpp [hunk 5]
```diff
@@ -1609,7 +1614,7 @@ void ClassDB::get_argument_options(const StringName &p_function, int p_idx, List
 				pf == "class_has_method" || pf == "class_get_method_list" ||
 				pf == "class_get_integer_constant_list" || pf == "class_has_integer_constant" || pf == "class_get_integer_constant" ||
 				pf == "class_has_enum" || pf == "class_get_enum_list" || pf == "class_get_enum_constants" || pf == "class_get_integer_constant_enum" ||
-				pf == "is_class_enabled" || pf == "is_class_enum_bitfield");
+				pf == "is_class_enabled" || pf == "is_class_enum_bitfield" || pf == "class_get_api_type" );
```
[the remark this answers is not in the document]
Comment (ZerxZ) [reply core/core_bind.cpp]: I don't understand what you're trying to say. What do I need to do? So I need to delete that part?

**Comment by ZerxZ** [comment]
> > so I just need to keep the `class_get_api_type` one?
> 
> I think yes, since the other two methods are already exposed.

I've removed that part and just kept the class_get_api_type part. Can you see if it's working? @dalexeev

File: doc/classes/ClassDB.xml [hunk 6]
```diff
@@ -31,6 +31,13 @@
 				Returns whether the specified [param class] is available or not.
 			</description>
 		</method>
+		<method name="class_get_api_type" qualifiers="const">
+			<return type="int" enum="ClassDB.APIType" />
+			<param index="0" name="class" type="StringName" />
+			<description>
+				Returns the api type of [param class]. See [enum APIType].
```
Comment (akien-mga) [inline doc/classes/ClassDB.xml:38]:
```suggestion
				Returns the API type of [param class]. See [enum APIType].
```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by ZerxZ** [comment]
> Thanks!
Thanks!

**Pull request merged**